File: 600.不含连续-1-的非负整数.py
#
# @lc app=leetcode.cn id=600 lang=python3
#
# [600] 不含连续1的非负整数
#

# @lc code=start
class Solution:
    def findIntegers(self, n: int) -> int:
        N = 31
        dp = [0] * N
        dp[0] = 1
        dp[1] = 1
        for i in range(2, N):
            dp[i] = dp[i - 1] + dp[i - 2]
        
        pre = 0
        ans = 0

        for i in range(N - 2, -1, -1):
            val = 1 << i
            if n & val:
                ans += dp[i + 1]
                if pre == 1:
                    break
                pre = 1
            else:
                pre = 0
            
            if i == 0:
                ans += 1
        return ans
# @lc code=end

File: 600.不含连续-1-的非负整数.cpp
/*
 * @lc app=leetcode.cn id=600 lang=cpp
 *
 * [600] 不含连续1的非负整数
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class Solution {
public:
    int findIntegers(int n) {
        vector<int> bits;
        while (n) {
            bits.push_back(n & 1);
            n >>= 1;
        }
        reverse(bits.begin(), bits.end());

        int m = bits.size(), dp[m + 1][2];
        memset(dp, -1, sizeof(dp));

        function<int(int, int, int)> dfs = [&](int i, int prev1, int isLimit) {
            if (i == m) return 1;
            if (!isLimit && dp[i][prev1] >= 0) return dp[i][prev1];
            int up = isLimit ? bits[i] : 1;
            int res = dfs(i + 1, 0, isLimit && up == 0);
            if (!prev1 && up == 1) {
                res += dfs(i + 1, 1, isLimit);
            }
            if (!isLimit) dp[i][prev1] = res;
            return res;
        };

        return dfs(0, 0, 1);
    }
};
// @lc code=end

